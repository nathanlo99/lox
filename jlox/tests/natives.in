5
1 1
2 5
3 4
69 420
5318008 1234567

10
There
once
was
a
man
named
Harry,
he
was
scary.
