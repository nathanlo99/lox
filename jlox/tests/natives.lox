
class String {
  init(str) { this.str = str; }
  get(idx) { return substr(this.str, idx, idx + 1); }
  substr(start, end) { return substr(this.str, start, end); }
  size { return length(this.str); }
}

var num_repeats = nextInt();
println("num_repeats = " + num_repeats);
for (var i = 0; i < num_repeats; i = i + 1) {
  var a = nextDouble();
  var b = nextDouble();
  if (a == nil)
    println("Got a = nil");
  else if (b == nil)
    println("Got b = nil");
  else
    println(a + b);
}

num_repeats = nextInt(); nextLine(); // Skip trailing newline char
for (var i = 0; i < num_repeats; i = i + 1) {
  var line = nextLine();
  var string = String(line);
  println("Line received: " + line);
  assert(string.str == line);
  if (length(line) > 3) {
  // println("substr(1, 3, line) is.. " + substr(1, 3, line)); // uh oh!
    println("substr(line, 1, 3) is.. " + string.substr(1, 3));
    assert(substr(line, 1, 3) == string.substr(1, 3));
  } else {
    println("Line too short...");
  }
}



var test_string = String("Hello, World!");
for (var i = 0; i < test_string.size; i = i + 1) {
  println("Index " + i + ": " + test_string.get(i));
}
