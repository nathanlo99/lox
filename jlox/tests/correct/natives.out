num_repeats = 5
2
7
7
489
6552575
Line received: There
substr(line, 1, 3) is.. he
Line received: once
substr(line, 1, 3) is.. nc
Line received: was
Line too short...
Line received: a
Line too short...
Line received: man
Line too short...
Line received: named
substr(line, 1, 3) is.. am
Line received: Harry,
substr(line, 1, 3) is.. ar
Line received: he
Line too short...
Line received: was
Line too short...
Line received: scary.
substr(line, 1, 3) is.. ca
Index 0: H
Index 1: e
Index 2: l
Index 3: l
Index 4: o
Index 5: ,
Index 6:  
Index 7: W
Index 8: o
Index 9: r
Index 10: l
Index 11: d
Index 12: !
